pkg/apigrpc/auth.go

```diff
@@ -0,0 +1,73 @@
+package apigrpc
+
+import (
+	"context"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// Authenticator exposes a function for authenticating requests.
+type Authenticator struct{}
+
+// Authenticate checks that a token exists and is valid. It stores the user
+// metadata in the returned context and removes the token from the context.
+func (a Authenticator) Authenticate(ctx context.Context) (context.Context, error) {
+	return a.tokenAuth(ctx)
+}
+
+const tokenPrefix = "Bearer "
+
+func (a Authenticator) tokenAuth(ctx context.Context) (context.Context, error) {
+	auth, err := extractAuthorization(ctx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !strings.HasPrefix(auth, tokenPrefix) {
+		return ctx, status.Error(codes.Unauthenticated, `missing "Bearer " prefix in "authorization" value`)
+	}
+
+	token := strings.TrimPrefix(auth, tokenPrefix)
+	if token == "" {
+		return ctx, status.Error(codes.Unauthenticated, "invalid token")
+	}
+
+	newCtx := purgeHeader(ctx, "authorization")
+
+	// TODO: hardcoded identity for now if token is not empty.
+	identity := &Identity{
```

Comment:
Here is a hardcoded Identity information about request issuer. If we aim to make this GRPC server generic then we should be able to handle requests from plugins and from users, properly authenticate and authorize them. In this case authorization decisions should leave inside each call implementation based on identity I suppose.